### PR TITLE
Example showing usage of resource from an external ROS package in Drake

### PR DESCRIPTION
**Pull request by azeey**
This adds an example of loading a URDF from an external ROS package. The URDF contains a mesh with a `package://` type URI. For this to be found, I populate the package map from `AMENT_PREFIX_PATH` as described in #170.

Closes  https://github.com/RobotLocomotion/drake-external-examples/issues/170

cc @IanTheEngineer , @EricCousineau-TRI 

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194)

### CODE REVIEW

File: drake_ament_cmake_installed/src/drake_example_pendulum/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# -*- mode: cmake -*-
+# vi: set ft=cmake :
+
+# Copyright (c) 2020, Massachusetts Institute of Technology.
```
Comment (jamiesnape) [inline drake_ament_cmake_installed/src/drake_example_pendulum/CMakeLists.txt:4]: 2021

File: drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE [hunk 2]
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2020, Massachusetts Institute of Technology.
```
Comment (jamiesnape) [inline drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE:1]: 2021

File: drake_ament_cmake_installed/src/drake_example_pendulum/package.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+  <name>drake_example_pendulum</name>
+  <version>0.1.0</version>
+  <description>Ament CMake project with an installed Drake</description>
```
Comment (jamiesnape) [inline drake_ament_cmake_installed/src/drake_example_pendulum/package.xml:6]: Can this be more specific?

File: drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 3.5)
```
Comment (jamiesnape) [inline drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt:1]: Can we stick with CMake 3.10 for consistency. And can you add the license block.

File: drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+project(drake_example_pendulum_description)
+
+find_package(ament_cmake REQUIRED)
```
Comment (jamiesnape) [inline drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt:4]: `find_package(ament_cmake CONFIG REQUIRED)`

**Comment by jwnimmer-tri** [comment]
My question on this one is at a higher level.

The "how to access a URDF" is properly within the scope of the build system, so is appropriate for this repository.  (That's usually the first hurdle where PRs fail in the concept stage, but in this case there is no problem.)

However, we still want to keep this repository minimal.  As of this PR we now have _two_ examples in the ament folder.  Do we need both particles and pendulum?  Can we show resource-loading as an add-on to the particle demo, for example, by using URDF help to render the scene?  Or can we remove the particle demo and only keep the pendulum?  Should we plan to add to the other (non-ament) folders a similar URDF demo, and should it be a pendulum in that case too, removing the particle everywhere?

I'd like to hear a statement and to have you all share thoughts with respect to the overall design intent and how this fits alongside everything else here.  The implementation details are boring and in pretty good shape already.

**Comment by jamiesnape** [comment]
From a build system demonstration point of view, we could lose the particle example, but it is our hello world that is implemented for all the other projects. If a Drake project is always likely to need this setup with resources, maybe it could replace the pendulum, and consistency is less of an issue.

**Comment by jamiesnape** [comment]
Ideally what whatever happens would be back ported to Drake somehow. If would like to be able to just generated the non-build system parts of this repro (offline) eventually. The code review should really happen in Drake itself.

**Review by IanTheEngineer — CHANGES_REQUESTED** [review]
From the perspective of a ROS2 user who wants to integrate Drake into their workflow, I think this example provides an easy entrypoint. Two packages, one a simple robot description with some (minimal) ROS dependencies, and the other being a Ament package that leverages drake but utilizes the external robot description directly. All this is to say, I think Addisu's packages function well as a standalone example.

With that said, I haven't looked into the particle example in much depth. It is entirely possible that we can reduce this folder to just two packages in the future. In that case, we would keep the description package (pendulum model) and then combine the particle and pendulum resource finding examples into one generic `drake_ament_examples` package, or something similarly named. That generic package would contain multiple executable examples.

As to the content of this PR, I think it looks great, keeping things minimal while covering how to find package resources from an Ament path. I left a few nits below to discuss.

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-:-MQshor69jZOGvd8MASf:b-w68w8)* status: 11 unresolved discussions, platform LGTM missing (waiting on @azeey)

___
*[drake_ament_cmake_installed/README.md, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsRj7H9IF8wXOFSVOG:-MQsRj7H9IF8wXOFSVOH:b-4m3die) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/README.md#L30)):*
> ```Markdown
> ```
> 
> To run the `drake_example_pendulum` example, source the workspace and run the executable
> ```

nit: line length for comments / docs in source is [generally limited to 80 char](https://drake.mit.edu/styleguide/cppguide.html#Line_Length). 

Also, end the line with a semicolon to present the code.
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsVivY3M7lMxf_xkEW:-MQsVivZ6csTUdEJzg_h:bqr4bkg) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE#L1)):*
> ```
> Copyright (c) 2020, Massachusetts Institute of Technology.
> ```

The [top-level repo license](https://github.com/RobotLocomotion/drake-external-examples/blob/master/LICENSE) has a copyright for both MIT and TRI. I think it makes sense to follow suit in this PR everywhere you invoke the license:

```
Copyright (c) 2021, Massachusetts Institute of Technology.
Copyright (c) 2021, Toyota Research Institute.
```
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/package.xml, line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsSVF_ESO_OyuEQ-uH:-MQsSVF_ESO_OyuEQ-uI:b-clvrs0) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/package.xml#L7)):*
> ```XML
>   <version>0.1.0</version>
>   <description>Ament CMake project with an installed Drake</description>
>   <maintainer email="addisu@openrobotics.org">Addisu Z. Taddese</maintainer>
> ```

nit: Unless you plan on maintaining this code, think the `<author>` tag makes more sense for you than the `<maintainer>` tag here :) 

However, as a [maintainer tag is required for package.xml format](http://wiki.ros.org/catkin/package.xml#Required_Tags), please feel free to add me as the maintainer:
```
<maintainer email="ian.mcmahon@tri.global"> Ian McMahon </maintainer>
```
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc, line 48 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsdta777AdU1bIOobd:-MQsdta777AdU1bIOobe:bwzdybx) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc#L48)):*
> ```
> // their resources, such as meshes
> ```
nit: `.` punctuation
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsdMZKDciEhQmFyQv_:-MQsdMZKDciEhQmFyQva:bx3m24q) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc#L56)):*
> ```
> could not be found. Have you sourced your work space?"
> ```
nit: `workspace`. I would also be more explicit about what you're referring to with workspace (as the definition of workspace is different in Drake/bazel lingo).
___
*[drake_ament_cmake_installed/src/drake_example_pendulum_description/package.xml, line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsXUE46vrONvcw8k-P:-MQsXUE5Es6PkYrzhqbM:btc62yc) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum_description/package.xml#L7)):*
> ```XML
>   <version>0.1.0</version>
>   <description>Example package to demonstrate using ROS resources in Drake</description>
>   <maintainer email="addisu@openrobotics.org">Addisu Z. Taddese</maintainer>
> ```

nit: ditto for the maintainer comments from the other package.xml

**Comment by jwnimmer-tri** [comment]
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/package.xml, line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsSVF_ESO_OyuEQ-uH:-MQsqzKNCpwUc6v4ddVX:b-lg0eua) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/package.xml#L7)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

nit: Unless you plan on maintaining this code, think the `<author>` tag makes more sense for you than the `<maintainer>` tag here :) 

However, as a [maintainer tag is required for package.xml format](http://wiki.ros.org/catkin/package.xml#Required_Tags), please feel free to add me as the maintainer:
```
<maintainer email="ian.mcmahon@tri.global"> Ian McMahon </maintainer>
```
</blockquote></details>

Actually Drake convention is to use lists, not people, for these:

https://github.com/RobotLocomotion/drake/blob/bd34d8cf15960145b31507747f41e084513c9e3a/manipulation/models/allegro_hand_description/package.xml#L11-L12

**Review by azeey — COMMENTED** [review]
I believe I've addressed all the code comments. Let me know if I should proceed with combining the particle and pendulum demos.

> Reviewed 10 of 10 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-:-MRQPn2jEDWiD8FL_8ws:b7otac)* status: 7 unresolved discussions, platform LGTM missing (waiting on @azeey, @IanTheEngineer, and @jamiesnape)

___
*[drake_ament_cmake_installed/README.md, line 30 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsRj7H9IF8wXOFSVOG:-MRQOKH06bz587fnbfjB:bybm7gc) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/README.md#L30)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

nit: line length for comments / docs in source is [generally limited to 80 char](https://drake.mit.edu/styleguide/cppguide.html#Line_Length). 

Also, end the line with a semicolon to present the code.
</blockquote></details>

Limited to 80 char and added colon in https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c.
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/CMakeLists.txt, line 4 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQmnCzs253z3oUKmwgE-r1-4:-MRQOSFJARvXXKDgiJ8P:b2cb4om) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/CMakeLists.txt#L4)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

2021
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQmnCzs253z3oUKmwgF-r1-1:-MRQOTzB-QourAVF2vQ4:b2cb4om) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE#L1)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

2021
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsVivY3M7lMxf_xkEW:-MRQOeYyBLBQC7Sm2yJO:b-98mux1) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/LICENSE#L1)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

The [top-level repo license](https://github.com/RobotLocomotion/drake-external-examples/blob/master/LICENSE) has a copyright for both MIT and TRI. I think it makes sense to follow suit in this PR everywhere you invoke the license:

```
Copyright (c) 2021, Massachusetts Institute of Technology.
Copyright (c) 2021, Toyota Research Institute.
```
</blockquote></details>

Done. I added TRI to all the files that invoke the license. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/package.xml, line 6 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQmnCzs253z3oUKmwgG-r1-6:-MRQOlqg30G--TZyUPqh:b1kgz94) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/package.xml#L6)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Can this be more specific?
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c.
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/package.xml, line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsSVF_ESO_OyuEQ-uH:-MRQOwWvByUJKVyPkTGU:byy394o) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/package.xml#L7)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Actually Drake convention is to use lists, not people, for these:

https://github.com/RobotLocomotion/drake/blob/bd34d8cf15960145b31507747f41e084513c9e3a/manipulation/models/allegro_hand_description/package.xml#L11-L12
</blockquote></details>

Done. I used the lists @jwnimmer-tri suggested. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc, line 48 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsdta777AdU1bIOobd:-MRQP2P1EdFFRVTZjDQd:b2cb4om) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc#L48)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

> ```
> // their resources, such as meshes
> ```
nit: `.` punctuation
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc, line 56 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsdMZKDciEhQmFyQv_:-MRQPBWY7eekhcff7B8x:bw9d167) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum/src/drake_example_pendulum.cc#L56)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

> ```
> could not be found. Have you sourced your work space?"
> ```
nit: `workspace`. I would also be more explicit about what you're referring to with workspace (as the definition of workspace is different in Drake/bazel lingo).
</blockquote></details>

I made it "ROS workspace" in https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c. How does that sound?
___
*[drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQmnCzs253z3oUKmwgI-r1-1:-MRQPIcz-I5NA9AzMPO8:b2cb4om) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt#L1)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Can we stick with CMake 3.10 for consistency. And can you add the license block.
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt, line 4 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQmnCzs253z3oUKmwgI-r1-4:-MRQPN4s1xsg1xiZEPXh:b2cb4om) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum_description/CMakeLists.txt#L4)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

`find_package(ament_cmake CONFIG REQUIRED)`
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c
___
*[drake_ament_cmake_installed/src/drake_example_pendulum_description/package.xml, line 7 at r1](https://reviewable.io/reviews/robotlocomotion/drake-external-examples/194#-MQsXUE46vrONvcw8k-P:-MRQPT0w94FSC3LZMxx_:b2cb4om) ([raw file](https://github.com/robotlocomotion/drake-external-examples/blob/2810ad5fe5a7a80ec509f89012235b2bd49b3e34/drake_ament_cmake_installed/src/drake_example_pendulum_description/package.xml#L7)):*
<details><summary><i>Previously, IanTheEngineer (Ian McMahon) wrote…</i></summary><blockquote>

nit: ditto for the maintainer comments from the other package.xml
</blockquote></details>

Done. https://github.com/robotlocomotion/drake-external-examples/commit/42fdbbcbd59549d439a5d852ccef95d9cf04199c

**Comment by jwnimmer-tri** [comment]
> It is entirely possible that we can reduce this folder to just two packages in the future.

I want to resolve the repository-design question now (i.e., have a plan), before we merge this.  If we think that amending the particle example to load resources is the correct plan, then we should not merge a pendulum in the meantime.  If we think having the pendulum be the sole example is the right plan, then this PR should remove the particle concurrently.  Given the relatively low attention and velocity we give to this repository, iterating and stumbling our way to a resolution will leave awkward intact for too long.

**Comment by IanTheEngineer** [comment]
This PR will be re-targeted toward the new `drake-ros` repository to exist as a standalone example.

**Pull request closed without merging**